R/standardize.data.R

```diff
@@ -282,6 +280,19 @@ standardize.grouped_df <- function(x,
 
 # helper -----------------------------
 
+.get_center_scale <- function(x, robust = FALSE, weights = NULL, reference = NULL) {
```

Comment:
The center / scale is now calculated in a separate function, for each variable, that takes the reference. 
Problem, that reference can only be another variable too (as the .get_center_scale helper is not able to query the "name" of a variable of a column so it cannot retrieve the appropriate column in a dataframe).


So in in standardize.data.frame, where we `lapply` standardize on each column, we need to pass the right "reference" (the right column of the dataframe corresponding to the column standardize is running on), and I don't know how to do that

Reply:
Something like this?

``` r
something <- function(x, select, ref = NULL) {
  f <- function(x, ref) {
    all(x == ref)
  }
  
  x <- x[select]
  ref <- if(!is.null(ref)) ref[select]
  
  mapply(x, if(!is.null(ref)) ref else x, 
         FUN = f)
}

something(mtcars[1:5,], select = c("mpg", "am"), mtcars[1:5,])
#>  mpg   am 
#> TRUE TRUE

something(mtcars[1:5,], select = c("mpg", "am"))
#>  mpg   am 
#> TRUE TRUE

something(mtcars[1:5,], select = c("mpg", "am"), mtcars[6:10,])
#>   mpg    am 
#> FALSE FALSE
```

<sup>Created on 2021-05-11 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

Reply:
@strengejacke 🤔

Reply:
must take a closer look at this, atm Idk what you're talking about... :-|

Reply:
I don't understand--why not just lapply() mean and sd across the selected variables, then pass those to the standardizing function and match by name?

Reply:
how so 🤔

Reply:
what is the `reference` for?

Reply:
Daniel be like

![](https://media.giphy.com/media/l3vRfPnPLHp2jFeCI/giphy.gif)

Reply:
lol it's to allow standardization based on the center/dispersion of another dataframe

Reply:
Ok, I still have no idea why one should use `reference`, but can't you use a named vector where `data[lhs]´ takes the reference of `reference[rhs]`? And if so, use `mapply()`.

Reply:
> or how it is supposed to work

Let's say you get or create some new data. And you want to see its "standardized score" relative to some population of reference. So you can do `standardize(newdata, reference = main_data)`

The intuitive solution would be to loop through each variable name, and then retrieve 1) the variable to standardize and 2) the parameters of standardization (centrality / dispersion) of the variable in the reference df and 3) apply the standardization.

But because we don't loop explicitly through the variable names, but use `lapply` to loop directly on the columns of the target df, we don't know how to tell at each iteration which column from the reference df to use for the reference.

Not sure it's clearer haha

Unfortunately I'm not good with `sapply/mapply/lapply/...` 😢

Reply:
This is rather quick and dirty, but when column names are identical, you could do this:

```r
test <- function(data, select = NULL, reference = NULL) {
  if (!is.null(select)) {
    data <- data[select]
  }

  cn <- colnames(data)

  if (!all(cn %in% colnames(reference))) {
    warning("Not all columns present in reference data.")
  }

  cn <- intersect(cn, colnames(reference))

  reference <- reference[cn]

  out <- as.data.frame(lapply(cn, function(i) {
    (data[[i]] - mean(reference[[i]], na.rm = TRUE)) / sd(reference[[i]], na.rm = TRUE)
  }))

  colnames(out) <- cn
  out
}

data(iris)
iris2 <- iris

test(iris, c("Sepal.Width", "Sepal.Length"), iris2)
```

Reply:
Yeah, I ended up replacing the lapply stuff with a loop.

RE grouped dfs, for now I put an error if a reference is set until we know how to deal with that.

@mattansb it's good on my side, feel free to review and merge